www/src/content/blog/2025-05-21-rate-limiting-hono-apps-intro.mdx

```diff
@@ -0,0 +1,291 @@
+
+### Connecting to remote databases
+
+Remote in-memory databases are the most common solution, as they minimize latency in the rate limit layer. Since some algorithms require multiple database calls, atomicity is another important consideration when choosing a storage solution.
+
+> Atomic operations are completed as a single unit, even if they have multiple steps. This prevents race conditions that could result in conflicting limits results. I’ll discuss this topic more in the next article on rate limiting algorithms.
+>
+
+[Redis](https://redis.io/docs) is a popular choice, both because it is extremely performant, and because it makes it easy to run multi-step operations atomically. Other in-memory databases like Memcached or Cloudflare KV can also be used, though they are geared towards caching and don’t offer robust support for atomic operations.
+
+Application architecture is arguably the most important factor though. Modern development patterns like distributed applications and edge functions introduce challenges and constraints that affect not just where data is stored, but also how it’s managed.
+
+### Distributed limiting
+
+While a centralized store makes rate limit records available across requests and app instances, it introduces a request-processing bottleneck. A flood of requests from instances around the world can bog down the database, causing even legitimate requests to lag or time-out.
+
+When paired with distributed application architecture, a centralized store can also add significant latency. After reaching the handler, each request must first be diverted to the limiter—possibly in a different region—before returning to the handling server for processing.
+
+In the era of edge functions (and runtimes), it’s no surprise that distributed rate limiters are increasingly popular. This strategy involves colocating a limiter data store with each app instance, thereby substantially reducing latency and avoiding a single point of failure.
+
+Of course, distributed rate limiters bring their own set of challenges. As with any distributed database, updates must be synchronized across instances. This ensures that users can’t bypass the limit simply by changing their “location” with a VPN. Special care must also be taken to mitigate race conditions, especially when using non-unique identifiers like IPs.
+
+### Data storage with `hono-rate-limiter`
+
+At the heart of `hono-rate-limiter` are it’s stores. These are adapters that implement the Fixed Window algorithm and communicate with the persistence layer. The library includes a Redis store compatible with a most Redis clients, as well as stores for Cloudflare [KV databases](https://developers.cloudflare.com/kv/) and [Durable Objects](https://developers.cloudflare.com/durable-objects/). It can also be paired with a [half-dozen third-party stores](https://github.com/@upstash/redis), and developers can build their own to connect with a different database or use an alternative algorithm.
```

Comment:
```suggestion
At the heart of `hono-rate-limiter` are its stores. These are adapters that implement the Fixed Window algorithm and communicate with the persistence layer. The library includes a Redis store compatible with a most Redis clients, as well as stores for Cloudflare [KV databases](https://developers.cloudflare.com/kv/) and [Durable Objects](https://developers.cloudflare.com/durable-objects/). It can also be paired with a [half-dozen third-party stores](https://github.com/@upstash/redis), and developers can build their own to connect with a different database or use an alternative algorithm.
```